docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,95 @@
+  * This behavior should be implemented as part of the base logic of both the HTTP and gRPC query APIs, before the query is handed to the query engine. This allows users managing complex Querier trees can choose where they want this logic to be enabled.
+* Update metrics exported by the components in the query path to include the tenant label when it's available.
+* Implement a tenant selector in the Query Frontend UI, which should communicate the tenant to Query Frontend using the HTTP header.
+
+<figure>
+<img height="80%" src="../img/query-path-tenancy-proposal-diagram.svg" width="80%"/>
+<figcaption>Diagram also available <a href="https://excalidraw.com/#json=ozCXvw8LBURpWkmiA4Loq,Nclh-Sw9g_mzO3gsM6yaRQ">online</a>.</figcaption>
+</figure>
+
+## Alternatives
+
+### Alternative implementations
+
+1. Apply verification and enforcement logic in the Query Frontend instead of Querier.
+
+The Query Frontend is an optional component on any Thanos deployment, while the Querier is always present. Plus, there might be deployments with multiple Querier layers where one or more might need to apply tenant verification and enforcement. On top of this, doing it in the Querier supports future work on using the [new Thanos PromQL engine](https://github.com/thanos-community/promql-engine), which can potentially make the Query Frontend unnecessary.
+
+2. Add the tenant identification as an optional field in the Store API protobuffer spec instead of an HTTP header.
+
+This isn't the preferred solution because the Store API isn't used all the way from the topmost components (e.g. Querier and Query Frontend) to the bottommost one (e.g. Store Gateway). The Query Frontend talks to the Querier via the Query API. A Querier can talk to other Queriers using the Query API via gRPC. Finally, a Querier talks to Store Gateways using the Store API. All these APIs would have to be modified and the transmission of this information would differ depending on the component and API being used. In addition to this complication, this implementation does not enable any capability that can't be achieved with HTTP header approach.
+
+### Alternative solutions
+
+1. Use a proxy to enforce and identify tenants in queries.
+
+While this could work for some of the features, like exporting per-tenant metrics, it would have to be inserted in front of many different components. Meanwhile, it doesn't solve the requirement for per-tenant configuration.
+
+2. Use a separate query path for each tenant.
+
+This incurs in a lot of wasted resources and demands manual work, unless a central tenant configuration is used and a controller is built around it to automatically manage the query paths.
+
+## Action Plan
+
+1. Query Frontend
+   1. Implement the `--query-frontend.tenant-header` (default is `""`") and `--query-frontend.default-tenant` (default is `"default-tenant"`) flags. Forward the tenant header to all the configured downstream query endpoints, if it is present.
```

Comment:
Configuring the tenant in multiple places can be a big source of confusion for users. Can query frontend simply propagate all headers? I would prefer it we enabled tenancy on the query path using one flag in one component.

Reply:
We need it to be configurable in multiple places because this header has to go all the way down to the Thanos Store, not only from Query Frontend to Query. It also has to be transmitted from Querier to Store Gateway in setups without the Query Frontend and from Querier to Querier in setups with multiple Querier trees. 

We need every component in the query path to be able to identify the header containing the tenant name, so there's no escape configuring this everywhere. If we configure it only in the Query Frontend, how does the Querier know which header includes the tenant's name?

I'm against doing a blanket forwarding of all headers. It seems like a potential source of security issues or unknown unintended consequences. WDYT?

Reply:
Basically, every component has to agree on which header is being used to identify the tenant and we have nothing to facilitate centralized configuration at the moment. 

Our only option is to hardcode the tenant header name to be always the same and impose it on all the users of Thanos that want to use this feature.

Reply:
Ah, there's a good reason for not hardcoding the tenant header's name: we don't do this in Receive. It seems better to me that we keep the same behavior in the query and write paths. Our default tenant header name is `THANOS-TENANT`.

Reply:
@fpetkovski it does work, but as I mentioned before: is not enough, not necessary, and breaks the pattern of the header that is used in Receive. 

It's not enough because: 

- Query Frontend has no awareness of the Store API at all (only the Queriers do, as they interact with Store Gateways), so how would I be able to observe it per-tenant and have per-tenant features there? It has to receive the tenant indication through the header or modifying the Query API. If we modify the Query API here, we won't be compatible with people simply pointing their Grafana datasources to a Query Frontend -- how would they add this extra "tenant" parameter to their requests? 
- Queriers can also receive requests via the Query API. So we also have to modify the Query API here or accept the header at this level too. 

It's not necessary because:

- A header can do the same thing, without any change to the Store API. Cortex and Mimir are proof that this works well. 
- A header works fine with Grafana datasource configurations.


The last point is self-explanatory: in our ingest path the tenancy is indicated through a (configurable) header on Receive. Why do it differently in the query path? 

So in the end, why would I prefer to modify the Query and Store APIs if using and propagating a single header is simpler while achieving the same features and covering every components with a single standard?

Reply:
Sorry but I feel this discussion is unfocused and unproductive. The original suggestion is to find a way to reduce the number of parameters that have to be configured to enable tenancy and I would appreciate it if we tried to work towards a solution together. My last comments were about the new `--store.tenant-header` which would be added to the store gateway because you preferred to keep the discussion in one thread.

Reply:
I agree with @fpetkovski here too. I don't think configuring parameters on every single Thanos component is the right way to go about this, and adding a field to StoreAPI would be way more simple. Let's work towards a solution on those lines. 

We already treat Ruler differently and out of scope for this proposal, we can do the same for Query frontend for the time being. Let's first ensure tenancy on the base Querier -> StoreAPI implementation flow, and then expand to other special components.

Reply:
> The original suggestion is to find a way to reduce the number of parameters that have to be configured to enable tenancy

@fpetkovski I understand your intention, but my point is that in doing so we can't achieve everything included in the proposal. We already have lots of configurations in every component, with some of them having to be repeated in more than one place, like replica label configuration, for example. Why pick on this now? Is one CLI flag in 3 components worth the complications in implementation? Can't we leave configuration improvement for a follow up proposal on centralized configuration, for example? This way we unblock the implementation of the features, that's the core intention of the proposal, and can work on improving not only this but a lot of configuration spread over the whole project.

To move forward with your (including @saswatamcode's) requested changes to the proposal, we would already have to exclude the Query Frontend and go directly to a solution that involves modifying Query API  and Store API. That's considerably more risk, because of it's harder to remove or change something in the Store API (it strives to be as stable as possible) and our Query API will start to diverge from Prometheus'. 

With your proposal in mind, I would ask to please give some thoughts and elaborate on the following questions that weren't answered yet. For us to work together towards something and have a productive discussions, I need to understand what would be the solutions in your head for these things:

* How will Querier forward tenant identification to subtrees of Queriers? The queries are sent using Prometheus Query API via gRPC. 
* How will Querier identify tenants in setups without the Query Frontend? Here the communication is done via Prometheus Query API.
* How will Querier identify tenants in setups with the Query Frontend? Do we really want to forward **all headers** from Query Frontend to Query? What about potential security impact and unintended consequences I mentioned before? If we decide it's fine to forward all headers, how will I have settings and observability per-tenant in Query Frontend?
* How could we update the Query Frontend later on to do the same?

Reply:
Needing 1, 3, or any arbitrary number of CLI flags to achieve a good implementation we are happy with as is a consequence and not the objective. I believe the focus should be on achieving a solution as simple, comprehensive, and standardized across components, as possible. I understand that the configuration of Thanos components is getting complex and repetitive. That's a problem on its own and it deserves some improvements, rather sooner than later I would say.

Reply:
@bwplotka me and @fpetkovski had a call this morning to better discuss. So it’s alright, no hard feelings between us. In fact we came to even better points than what we talked about here. 😄

Either me or him will post the summary of the conversation later. We found a good middle point regarding both approaches to ensure compatibility with other tools (like Cortex and others), to have consistent behavior across components (how Receive behaves regarding its tenant header), minimize amount of config flags, and avoid modifying APIs.

Reply:
We synced up yesterday via video call and agreed that tenant propagation can happen via a pre-defined header or field in the `SeriesRequest` message, which should remove the need for a flags in store components. 

There is still no good way to have per-tenant features in query frontend, so we might need to have a flag in both the querier and QFE if we want tenancy at the frontend level.

Reply:
Actually now that I think about this, label injection would happen in the Querier, so only the Querier needs a flag to indicate the tenant label. If we go with @bwplotka's proposal to use pre-defined header names, Query Frontend just needs to forward the header to the Querier, so we won't need additional configuration options.

Reply:
@fpetkovski as we spoke yesterday, I think we **need** customizable header name because Receive's already offering that (through `--receive.tenant-header="THANOS-TENANT"`) and using two different header names depending on whether it's a query or a remote write creates confusion through inconsistency. 

I think it's valuable to keep the consistency here so that people that already know how Receive works will easily understand how the tenancy enforcement on the query path works.

The default header name will also match Receive's. So we will end up with:

  * `--querier.tenant-header="THANOS-TENANT"`. 
  * `--querier.default-tenant="default-tenant"`.
  * `--querier.tenancy="false"`.

When using the default header name, you will have to only add `--querier.tenancy="true"` to toggle everything on. The simplest case is gracefully simple to handle.

And if you are using a custom header name in Receive, you may (or may not) make the query path consistent with it by customizing `--querier.tenant-header`. The same applies to the default tenant setting.

cc @bwplotka

Reply:
As also mentioned by @bwplotka, the possibility of customizing the header name allows us to be more compatible with other tools, like Cortex, which facilitates interoperation and migration between tools.

Reply:
That makes sense to me.

Reply:
@fpetkovski actually, we need that header name flag in the Query Frontend too. It cannot simply forward all headers. It needs to know what is the tenant header to include the information in its own metrics and for the per-tenant settings, as previously mentioned in our conversations.